src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/AsyncCausalityTracer.cs

```diff
@@ -0,0 +1,105 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace System.Threading.Tasks
+{
+    internal static class AsyncCausalityTracer
+    {
+        internal static void Enable()
+        {
+            Interlocked.Increment(ref listenerCnt);
+        }
+
+        internal static void Disable()
+        {
+            Interlocked.Decrement(ref listenerCnt);
+        }
+
+        internal static bool LoggingOn => listenerCnt > 0;
+        private static int listenerCnt = 0;
+
+        // The TraceXXX methods should be called only if LoggingOn property returned true
+        //
+        [MethodImpl(MethodImplOptions.NoInlining)] // Tracking is slow path. Disable inlining for it.
+        internal static void TraceOperationCreation(Task task, string operationName)
+        {
+            try
+            {
+                if (LoggingOn)
```

Comment:
Why do we need to duplicate the check here when the above says the caller should check it?